### ISSUE TITLE
Sometimes upgrade to websocket is unsuccessful and blocks communication

### ISSUE THREAD

**Issue by ericremoreynolds** [miguelgrinberg/Flask-SocketIO#144]
Hi Miguel,

We are attempting to substitute the old flask-socketio with the new one in an existing project, but we are having difficulties getting it to work. The erratic behaviour can be reproduced in your sample app as follows. Try substituting

```
socketio = SocketIO(app, async_mode=async_mode)
```

with

```
socketio = SocketIO()
socketio.init_app(app, async_mode=async_mode)
```

Our understanding based on past experience of Flask extensions is that these two alternatives should do the same thing.

However you should notice in the sample app (with an async mode which supports websockets) that engineio logs 'received request to upgrade' but not 'successfully upgraded', and subsequent messages sent by the client (apparently successfully) are not received by the server.

Can you reproduce this behaviour? If so can you confirm thats it's a bug or if we have misunderstood how to use the new extension?

Thanks in advance,

Kind regards,

Eric

**Comment by ericremoreynolds** [comment]
Actually on further investigation it seems that the method of initialization is not the discriminating factor (looking at the code in fact it really cannot be). I have now noticed that with the sample app, in an unpredictable way sometimes the connection does not get successfully upgraded and when this happens subsequent messages from the client are ignored by the server.

**Comment by miguelgrinberg** [comment]
Can you share the output of `pip freeze`? I want to try to reproduce with the same virtualenv you have.

**Comment by ericremoreynolds** [comment]
alabaster==0.7.6
alembic==0.7.6
Babel==2.1.1
blinker==1.4
boto==2.38.0
docutils==0.12
eventlet==0.17.4
Flask==0.10.1
Flask-Admin==1.3.0
Flask-DebugToolbar==0.10.0
Flask-Login==0.2.11
Flask-Mail==0.9.1
Flask-SocketIO==1.0b1
Flask-SQLAlchemy==2.0
Flask-WTF==0.11
geoip2==2.2.0
gevent==1.0.2
gevent-websocket==0.9.5
greenlet==0.4.7
gunicorn==18.0
IbPy==0
ipaddr==2.1.11
itsdangerous==0.24
Jinja2==2.7.3
Mako==1.0.2
MarkupSafe==0.23
maxminddb==1.2.0
mock==1.0.1
mockredispy==2.9.0.11
nose==1.3.7
Pygments==2.0.2
pyjade==3.0.0
python-engineio==0.6.7
python-socketio==0.5.2
pytz==2015.6
redis==2.10.3
requests==2.7.0
selenium==2.46.0
six==1.9.0
snowballstemmer==1.2.0
Sphinx==1.3.1
sphinx-rtd-theme==0.1.9
SQLAlchemy==1.0.5
tzlocal==1.1.2
ujson==1.33
WebSmsComToolkit==1.0.2
Werkzeug==0.10.4
wheel==0.26.0
wsaccel==0.6.2
WTForms==2.0.2

Rgds

**Comment by ericremoreynolds** [comment]
Ah I don't know if this could be relevant: I'm using Miniconda and using conda env to create the environment.

**Comment by miguelgrinberg** [comment]
Which version of Python are you using through conda?

**Comment by ericremoreynolds** [comment]
Python 2.7.10 |Continuum Analytics, Inc.| (default, Sep 15 2015, 14:29:08) 
[GCC 4.2.1 (Apple Inc. build 5577)] on darwin

Mac OS X 10.10.3 though I have noticed similar (if not identical) on other operating systems (Windows) with a similar Python setup.

**Comment by ktoh** [comment]
@ericremoreynolds i just did an upgrade from 0.6 to 1.0 and ran into the same problem. 

In my case, I found that the issue was client side, not server side:

* Prior to v1.0b I was using socketio 0.9.16 as recommended by the documentation (http://flask-socketio.readthedocs.org/en/latest/) 

* After the upgrade, I found that 0.9.16 didn't connect successfully with the server configured to run with eventlets (`async_mode=="eventlet"`).   However, the current  release of socketio client (1.3.7 today) worked fine.

**Comment by miguelgrinberg** [comment]
@ktoh Ah, that's a very good point. Looks like I need to write some docs on how to upgrade from the 0.6x to the 1.0x releases.

**Comment by ktoh** [comment]
@miguelgrinberg the update went very smoothly. Frankly it came at a great time because there are some real problems with older gevent issues like this ugliness:

https://github.com/gevent/gevent/issues/477

I'd offer to write a migration FAQ but so far there aren't really any issues to document

**Comment by miguelgrinberg** [comment]
@ktoh Yes, I also like eventlet much better. And WebSocket performs much better in eventlet than in gevent.

As far as the docs, I'll just add a short section with a check list of things that you need to do for the upgrade. Things such as upgrading the socket.io client, uninstalling gevent-socketio, etc.

**Comment by ktoh** [comment]
@miguelgrinberg let me know if u want help on that, happy to pitch in

**Comment by ericremoreynolds** [comment]
@ktoh Thanks for your input but I don't think that is my issue here. I am using 1.3.5 and like I said I am encountering these problems even with the sample app.

When I get some time I will post a screencast.

**Comment by ktoh** [comment]
@ericremoreynolds got it. The issue I ran into was also evident in the sample app (that's how I figured out it was client side), but your issue sounds different.

**Comment by ericremoreynolds** [comment]
https://youtu.be/SeXDfrcYk78

**Comment by miguelgrinberg** [comment]
@ericremoreynolds Very nice, good call on enabling logs on the screencast. The timing of events in your log around the websocket upgrade is slightly different than what I see here in mine. That gives me something to investigate. Thanks, I'll keep you posted.

**Comment by miguelgrinberg** [comment]
@ericremoreynolds I was able to reproduce this on a Mac, luckily. The fix is in package python-engineio v0.7.1. Could you please upgrade to the following versions and check if the problem is also gone for you?

    Flask-SocketIO==1.0b1
    python-socketio==0.6.0
    python-engineio==0.7.1

**Comment by ericremoreynolds** [comment]
Hi Miguel,
I confirm that with the latest versions of the three libraries it's working for me too.
Thanks very much!
